### ISSUE TITLE
In REST API Snapshot options are quite confusing. 

### ISSUE THREAD

**Issue by Shrivaibavi** [gluster/glusterd2#1165]
Description:
When trying to create a snapshot by using REST API, The options that i was tend to use was quite confusing.
curl -X GET http://10.70.35.5:24007/v1/snapshots
curl -X POST http://10.70.35.5:24007/v1/snapshot --data @snap.json -H 'Content-Type: application/json'

Using v1/snapshots for getting the list and using v1/snapshot for creating snapshot is quite confusing.
The GET and POST options can be made similar, Either v1/snapshot or v1/snapshots.
The problem here is when i used this command  "curl -X POST http://10.70.35.5:24007/v1/snapshots --data @snap.json -H 'Content-Type: application/json' " the curl command returns nothing and the user will not be sure whether the snapshot got created or not.

**Comment by rafikc30** [comment]
the command `curl -X POST http://10.70.35.5:24007/v1/snapshots` lists all snapshots or snapshots of a volume, but `curl -X POST http://10.70.35.5:24007/v1/snapshot --data @snap.json -H 'Content-Type: application/json'` creates a single snapshot. That is why it is not identical.

But you should get an error when using `curl -X POST http://10.70.35.5:24007/v1/snapshots  --data @snap.json -H 'Content-Type: application/json'` ,  because there are rest endpoints defined with that name. Can you please verify it.

**Comment by Shrivaibavi** [comment]
@rafikc30 I tried as you said, I am not getting any error when i am using /v1/snapshots.


[root@dhcp35-5 ~]# curl -X POST http://10.70.35.5:24007/v1/snapshots --data @snap.json -H 'Content-Type: application/json'
[root@dhcp35-5 ~]# glustercli snapshot list
+------+---------------+
| NAME | ORIGIN VOLUME |
+------+---------------+
+------+---------------+
[root@dhcp35-5 ~]# curl -X POST http://10.70.35.5:24007/v1/snapshot --data @snap.json -H 'Content-Type: application/json'
{"snapinfo":{"id":"7f1c8d2a-4aba-4846-ac2d-f89d7c2b2b41","name":"snap3","type":"Replicate","transport":"tcp","distribute-count":1,"replica-count":3,"options":{"bitrot-stub.bitrot":"off","feature.deem-statfs":"off","features.barrier":"disable","features.inode-quota":"off","features.quota":"off","features.quota-deem-statfs":"off","features.read-only":"on","features.uss":"off","replicate.self-heal-daemon":"off"},"state":"Created","subvols":[{"name":"snap3-replicate-0","type":"Replicate","bricks":[{"id":"71f070ec-6eae-4757-86f6-e2941fc3163c","path":"/var/run/glusterd2/snaps/vol2/d8750991ab2749ca92910897a0c7a766/snap3/brick1/vol2","volume-id":"7f1c8d2a-4aba-4846-ac2d-f89d7c2b2b41","volume-name":"snap3","peer-id":"8fa3e7e6-a5ac-47b2-9ba3-ca29997dbeb0","host":"10.70.35.5","type":"Brick"},{"id":"c7ba9a6b-0cdf-4d6f-b391-26ca90aec3e5","path":"/var/run/glusterd2/snaps/vol2/24819888487f4d43b9b8f23362a3d4ee/snap3/brick1/vol2","volume-id":"7f1c8d2a-4aba-4846-ac2d-f89d7c2b2b41","volume-name":"snap3","peer-id":"d112190f-5213-417f-83c6-4ba045ead1c4","host":"10.70.35.47","type":"Brick"},{"id":"63d5c3ab-1ab0-483f-b0e4-b243d9345419","path":"/var/run/glusterd2/snaps/vol2/25cf9ac94757494dbbb8ee188bf7df87/snap3/brick1/vol2","volume-id":"7f1c8d2a-4aba-4846-ac2d-f89d7c2b2b41","volume-name":"snap3","peer-id":"3c23daaf-f343-4ac8-bd3f-186a44101df2","host":"10.70.35.179","type":"Brick"}],"replica-count":3,"arbiter-count":0,"disperse-count":0}],"metadata":null,"snap-list":null},"parentname":"vol2","description":""}
[root@dhcp35-5 ~]# glustercli snapshot list
+-------+---------------+
| NAME  | ORIGIN VOLUME |
+-------+---------------+
| snap3 | vol2          |
+-------+---------------+

**Comment by rafikc30** [comment]
I think the problem is with package mux. When there is a mismatch in method it write the status as
`// methodNotAllowed replies to the request with an HTTP status code 405.
func methodNotAllowed(w http.ResponseWriter, r *http.Request) {
        w.WriteHeader(http.StatusMethodNotAllowed)
}
`
rather if it was as below, then result would have been as expected
`// methodNotAllowed replies to the request with an HTTP status code 405.
func methodNotAllowed(w http.ResponseWriter, r *http.Request) {
        http.Error(w, "405 method not allowed", http.StatusMethodNotAllowed)
       w.WriteHeader()
}
`
So the point is this is not a snapshot bug, if you try for other http routes you will get the same result
for example 
`curl -X POST http://`hostname`:24007/v1/events` also return nothing, because it allows method GET.
So I request you to close this issue

**Comment by rishubhjain** [comment]
@rafikc30 yes, I have filed an issue in gorilla/mux as well to get more details about this issue. This issue   is for all the URLs having just the GET request or the POST request, I think we will have to handle the other cases as well. May be this issue can be kept open and marked as a general bug instead of snapshot specific bug.

**Comment by aravindavk** [comment]
I am seeing multiple issues here.

- No uniformity in Snapshots URL similar to volumes URL, we should use `/snapshots` as prefix for all snapshot related commands.  For example, snapshot create URL should be `POST /v1/snapshots` and Snapshot activate `POST /v1/snapshots/:snapname/activate`
- If we are using `POST` for getting some information then it is wrong(As mentioned here https://github.com/gluster/glusterd2/issues/1165#issuecomment-412498912)

**Comment by aravindavk** [comment]
> @rafikc30 yes, I have filed an issue in gorilla/mux as well to get more details about this issue. This issue is for all the URLs having just the GET request or the POST request, I think we will have to handle the other cases as well. May be this issue can be kept open and marked as a general bug instead of snapshot specific bug.

I don't think it is `gorilla/mux` issue. If a URL doesn't accept a method then it should return the `http.StatusMethodNotAllowed`

**Comment by atinmu** [comment]
> No uniformity in Snapshots URL similar to volumes URL, we should use /snapshots as prefix for all snapshot related commands. For example, snapshot create URL should be POST /v1/snapshots and Snapshot activate POST /v1/snapshots/:snapname/activate
> If we are using POST for getting some information then it is wrong(As mentioned here #1165 (comment))

Can we please review all the snapshot ReST APIs in detail and understand the gaps and get them fixed at earliest?

**Comment by rishubhjain** [comment]
@Shrivaibavi Try `curl -XPOST -sv http://10.70.35.5:24007/v1/snapshots --data @snap.json -H 'Content-Type: application/json` you will get a 405 Method Not Found error.

**Comment by rafikc30** [comment]
>I don't think it is gorilla/mux issue. If a URL doesn't accept a method then it should return the http.StatusMethodNotAllowed

I think, the error http.StatusMethodNotAllowed  is treating different from http.NotFound from gorilla/mux, as mentioned in comment https://github.com/gluster/glusterd2/issues/1165#issuecomment-416667121 they are not sending this as an error. We have similiar problem in all endpoints, for example the "/v1/event" rest endpoint has similiar issue. As @rishubhjain pointed out, We could see the correct error message if we include `curl -sv` to the request. I'm not sure the inention behind this behaviour. But I believe we don't need to do anything from our end to fix this issue.

>If we are using POST for getting some information then it is wrong(As mentioned here #1165 (comment))

We are not using post to get any information, this issue is about not throwing error when using an undefined method. But this is not specific to snapshot comments.

>No uniformity in Snapshots URL similar to volumes URL, we should use /snapshots as prefix for all snapshot related commands. For example, snapshot create URL should be POST /v1/snapshots and Snapshot activate POST /v1/snapshots/:snapname/activate

 We can make everything to a unifrom prefix. As mentioned in comment https://github.com/gluster/glusterd2/issues/1165#issuecomment-412498912 , my interntion was to use `snaphot` for single entity and `snapshots` for multiple entity.

**Comment by rafikc30** [comment]
This issue is solved with PR https://github.com/gluster/glusterd2/pull/1213. @Shrivaibavi  Can this be closed now?
